snet_cli/mpe_client_command.py

```diff
@@ -12,61 +12,186 @@
+    def _init_channel_from_metadata(self, channel_dir, metadata):
+        self._check_mpe_address_metadata(metadata)
+        if (os.path.exists(channel_dir)):
+            raise Exception("Directory %s already exists"%channel_dir)
+        os.makedirs(channel_dir)
+        try:
+            spec_dir = os.path.join(channel_dir, "service_spec")
+            os.makedirs(spec_dir)
+            # take tar of .proto files from ipfs and extract them to channel_dir/service_spec
+            spec_tar = get_from_ipfs_and_checkhash(self._get_ipfs_client(), metadata["model_ipfs_hash"])
+            with tarfile.open(fileobj=io.BytesIO(spec_tar)) as f:
+                f.extractall(spec_dir)
+
+            # compile .proto files
+            if (not compile_proto(Path(spec_dir), channel_dir)):
+                raise Exception("Fail to compile %s/*.proto"%spec_dir)
+
+            # save service_metadata.json in channel_dir
+            metadata.save(os.path.join(channel_dir, "service_metadata.json"))
+
+            # save current ethereum address in channel dir
+            with open(os.path.join(channel_dir, "client_address.txt"), 'w') as f:
+                f.write(self.ident.address)
+        except:
+            # it is secure to remove channel_dir, because we've created it
+            shutil.rmtree(channel_dir)
+            raise
+
+    def _check_channel_is_mine(self, channel_id):
+        channel = self._get_channel_state_from_blockchain(channel_id)
+        if (channel["sender"].lower() != self.ident.address.lower() and
+            channel["signer"].lower() != self.ident.address.lower()):
+                raise Exception("Channel %i does not correspond to the current Ethereum identity "%channel_id + 
+                                "(address=%s sender=%s signer=%s)"%(self.ident.address.lower(), channel["sender"].lower(), channel["signer"].lower()))
+
+    def init_channel_from_metadata(self):
+        metadata  = load_mpe_service_metadata(self.args.metadata_file)
+        self._check_channel_is_mine(self.args.channel_id)
+        self._init_channel_from_metadata(self.get_channel_dir(), metadata)
+
+    def init_channel_from_registry(self):
+        channel_dir   = self.get_channel_dir()
+        metadata      = self._get_service_metadata_from_registry()
+        self._check_channel_is_mine(self.args.channel_id)
+        self._init_channel_from_metadata(channel_dir, metadata)
+
+    def _open_channel_for_service(self, metadata):
+        group_id    = metadata.get_group_id(self.args.group_name)
+        recipient   = metadata.get_payment_address(self.args.group_name)
+        mpe_address = self.get_mpe_address()
+        rez = self.transact_contract_command("MultiPartyEscrow", mpe_address, "openChannel", [recipient, self.args.amount, self.args.expiration, group_id, self.ident.address])
+
+        if (len(rez[1]) != 1 or rez[1][0]["event"] != "ChannelOpen"):
+            raise Exception("We've expected only one ChannelOpen event after openChannel")
```

Comment:
It is a sanity check. Normally it could happen only if MPE contract was changed (or something broken somewhere). I could replace error message with "We've expected only one ChannelOpen event after openChannel. Make sure that you use correct MultiPartyEscrow address. "